aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -799,4 +800,241 @@ Tensor sspaddmm(const Tensor& self, const Tensor& mat1, const Tensor& mat2,
+  return at::_sparse_sum(input.to(dtype), dims_to_sum, false);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, bool keepdim) {
+
+  AT_CHECK(input.is_coalesced(), "To support autograd, input SparseTensor has to be coalesced.");
+
+  const int64_t input_dim = input.dim();
+  dim_list_to_bitset(dims_to_sum, input_dim);
+  auto dims_to_sum_v = dims_to_sum.vec();
+  maybe_wrap_dims(dims_to_sum_v, input_dim);
+
+  std::bitset<64> dims_to_sum_b;
+  for (auto d : dims_to_sum_v) {
+    dims_to_sum_b[d] = true;
+  }
+
+  LongTensor indices = input._indices();
+  Tensor values = input._values();
+  IntList sizes = input.sizes();
+  const int64_t sparse_dim = input.sparse_dim();
+  const int64_t dense_dim = input.dense_dim();
+
+  int64_t sparse_dims_to_sum_size = 0;
+  auto dims_to_keep_v = std::vector<int64_t>();
+  Tensor sparse_dims_to_keep_t = at::zeros({sparse_dim}, indices.options().dtype(at::kByte));
+  auto dense_dims_to_sum_v = std::vector<int64_t>();
+  for (int64_t d = 0; d < input_dim; d++) {
+    if (dims_to_sum_b[d]) {
+      if (d < sparse_dim) sparse_dims_to_sum_size ++;
+      else dense_dims_to_sum_v.emplace_back(d + 1 - sparse_dim);
+    }
+    else {
+      dims_to_keep_v.emplace_back(d);
+      if (d < sparse_dim) sparse_dims_to_keep_t[d].fill_(1);
+    }
+  }
+
+  const bool sum_all_sparse_dim = (sparse_dim == sparse_dims_to_sum_size);
+
+  // new indices
+  LongTensor new_indices;
+  if (keepdim) {
+    new_indices = indices.clone();
+    if (sparse_dims_to_sum_size > 0) {
+      for (auto d : dims_to_sum_v) {
+        if (d < sparse_dim) new_indices[d].fill_(0);
+      }
+    }
+  }
+  else {
+    if (sum_all_sparse_dim) {
+      new_indices = at::zeros({1, input._nnz()}, indices.options());
+    }
+    else if (sparse_dims_to_sum_size > 0) {
+      new_indices = indices.index_select(0, at::arange(sparse_dim, indices.options()).masked_select(sparse_dims_to_keep_t));
+    }
+    else {
+      new_indices = indices.clone();
+    }
+  }
+
+  // new size
+  std::vector<int64_t> new_sizes;
+  if (keepdim) {
+    new_sizes = sizes.vec();
+    for (auto d : dims_to_sum) new_sizes[d] = 1;
+  }
+  else {
+    for (auto d : dims_to_keep_v) new_sizes.emplace_back(sizes[d]);
+    if (sum_all_sparse_dim) new_sizes.emplace(new_sizes.begin(), 1);
+  }
+
+  // new values
+  Tensor new_values = values.clone();
+  if (dense_dims_to_sum_v.size() > 0) {
+    new_values = new_values.sum(dense_dims_to_sum_v);
+
+    if (keepdim) {
+      auto new_values_size = values.sizes().vec();
+      for (auto d : dense_dims_to_sum_v) new_values_size[d] = 1;
+      new_values = new_values.reshape(new_values_size);
+    }
+  }
+
+  SparseTensor new_sparse = at::sparse_coo_tensor(new_indices.contiguous(), new_values.contiguous(), new_sizes, input.options());
+  new_sparse = new_sparse.coalesce();
+
+  if (sum_all_sparse_dim) {
+    // return a dense tensor if sum over all sparse dims
```

Comment:
:O Can you have a sparse tensor with zero sparse dimensions? If so, when should be used, over a dense tensor?

Reply:
It would be reasonable to make a blanket statement about sparse tensors: "any time you would have a sparse tensor with zero dims, you should turn it into a dense tensor." We should document this prominently somewhere, probably the top of the sparse documentation.

It's unfortunate that this doesn't fall out automatically from our data representation.